DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky/src/cholesky.hpp

```diff
@@ -0,0 +1,159 @@
+                            AMatrixPipe>(a_device, matrix_count, repetitions);
+  });
+
+  // Read the A matrix from the AMatrixPipe pipe and compute the Cholesky
+  // decomposition. Write the L output matrix to the LMatrixPipe pipe.
+  q.single_task<Cholesky>(
+      fpga_linalg::StreamingCholesky<T, is_complex, dimension, raw_latency,
+                        kNumElementsPerDDRBurst, AMatrixPipe, LMatrixPipe>());
+
+  auto ddr_write_event =
+  q.single_task<CholeskyLocalMemToDDRL>([=]() [[intel::kernel_args_restrict]] {
+    // Read the L matrix from the LMatrixPipe pipe and copy it to the FPGA DDR
+    // Number of DDR bursts of kNumElementsPerDDRBurst required to write
+    // one vector
+    constexpr bool kIncompleteBurst =
+       kLMatrixSize % kNumElementsPerDDRBurst != 0;
+    constexpr int kExtraIteration = kIncompleteBurst ? 1 : 0;
+    constexpr int kLoopIter =
+       (kLMatrixSize / kNumElementsPerDDRBurst) + kExtraIteration;
+
+    sycl::device_ptr<TT> vector_ptr_device(l_device);
+
+    // Repeat matrix_count complete L matrix pipe reads
+    // for as many repetitions as needed
+    [[intel::loop_coalesce(2)]]
+    for (int rep_idx = 0; rep_idx < repetitions; rep_idx++) {
+      for (int matrix_idx = 0; matrix_idx < matrix_count; matrix_idx++) {
+        for (int li = 0; li < kLoopIter; li++) {
+          TT bank[kNumElementsPerDDRBurst];
+
+          for (int k = 0; k < kNumElementsPerDDRBurst; k++) {
+            if(((li * kNumElementsPerDDRBurst) + k) < kLMatrixSize){
+              bank[k] = LMatrixPipe::read();
+            }
+          }
+
+          // Copy the L matrix result to DDR
+          if constexpr (kIncompleteBurst){
+            // Write a burst of kNumElementsPerDDRBurst elements to DDR
+            #pragma unroll
+            for (int k = 0; k < kNumElementsPerDDRBurst; k++) {
+              if(((li * kNumElementsPerDDRBurst) + k) < kLMatrixSize){
+                vector_ptr_device[(matrix_idx * kLMatrixSize)
+                          + (li * kNumElementsPerDDRBurst) + k] = bank[k];
+              }
+            }
+          }
+          else{
+            // Write a burst of kNumElementsPerDDRBurst elements to DDR
+            #pragma unroll
+            for (int k = 0; k < kNumElementsPerDDRBurst; k++) {
+              vector_ptr_device[(matrix_idx * kLMatrixSize)
+                          + (li * kNumElementsPerDDRBurst) + k] = bank[k];
+            }
+          }
+        }  // end of li
+      }  // end of matrix_idx
+    }    // end of rep_idx
+  });
+
+  ddr_write_event.wait();
+
+  // Compute the total time the execution lasted
+  auto start_time = ddr_read_event.template
+              get_profiling_info<sycl::info::event_profiling::command_start>();
+  auto end_time = ddr_write_event.template
+                get_profiling_info<sycl::info::event_profiling::command_end>();
+  double diff = (end_time - start_time) / 1.0e9;
+  q.throw_asynchronous();
```

Comment:
The throw_async is OK here, but it might help to document why you are calling it here and what errors you might be trying to catch.   Otherwise it looks like random placement.